rfd/0095-application-access-okta-integration.md

```diff
@@ -0,0 +1,315 @@
+teleport start --okta-api-token 12345678
+
+# Environment variable example
+TELEPORT_OKTA_API_TOKEN="12345678" teleport start
+```
+
+### Okta user traits
+
+Okta will optionally use an `okta_user_id` trait to map Teleport users to Okta users if necessary.
+
+```yaml
+kind: user
+version: v2
+metadata:
+  name: alice
+spec:
+  roles: ['devs']
+  traits:
+    logins: ['admin']
+    kubernetes_groups: ['edit']
+    okta_user_id: ['name@mydomain.com']
+```
+
+If Okta is used for logging into Teleport or the Teleport user's username is the same as the
+e-mail used for Okta, this will be unnecessary.
+
+### Background synchronization
+
+The background synchronization process, which synchronizes all applications from Okta, is expected
+to run roughly every 2 minutes. This process wil translate Okta users, groups, and applications
+into their Teleport equivalents.
+
+```mermaid
+sequenceDiagram
+  participant OktaSvc as Okta Service
+  participant Okta as Okta API
+  participant AppSvc as Application Access Service
+
+  loop Every 2 minutes
+    OktaSvc->>Okta: Get list of applications
+    OktaSvc->>AppSvc: Synchronize applications in Teleport
+
+    OktaSvc->>Okta: Get list of groups
+
+    OktaSvc->>AppSvc: Synchronize groups in Teleport
+  end
+```
+
+#### Okta to Teleport mappings
+
+Okta groups and applications will be mapped by the background synchronization into Teleport roles and
+applications respectively.
+
+##### Groups
+
+A Teleport role will be created for each Okta group. Each group will be prepended with a configurable
+`role_prefix` which defaults to `okta/` and then suffixed with the type and name of the object, delimited by
+`/` characters.  For example, if an Okta group called `Developers` was synchronized to Teleport,
+it would be called `okta/group/Developers`.
+
+These roles use a configurable `role_template` when created which defaults to allowing users access to
+applications that belong to this group. The resulting role for `okta/group/Developers` would look like:
+
+```yaml
+kind: role
+version: v5
+metadata:
+  name: okta/group/Developers
+spec:
+  allow:
+    app_labels:
+      group: "okta/group/Developers"
+```
+
+The corresponding application labels will be populated by application synchronization.
+
+##### Applications
+
+When applications are synchronized with Teleport, they will be created in application access as
+HTTP apps that use the `appLinks` from Okta as their URI. If there is more than one `appLink`
+associated with an Okta application, it will be split into multiple applications for
+each `appLink` with the unique name of each `appLink` used to disambiguate them. The `origin` field
+in the application metadata will be set to `okta`.
+
+Applications in Okta can be assigned to groups, and the groups that the application is assigned
+to will be used to create labels for the application.
+
+One complicating factor for us is that Okta allows per-application user assignment as well. As a
```

Comment:
Why don't we extend our Role resource to support access control to specific apps/groups instead of creating roles and using app labels? For example:

```yaml
kind: role
version: v5
metadata:
  name: example
spec:
  allow:
    okta_apps: ["app1", "app2"]
    okta_groups: ["group1", "group2"]
```

Reply:
This makes sense. I'll adjust!